File: examples/phase/c++/simple.cpp
#include <stdio.h>
#include <unistd.h>
#include <TAU.h>


int f1(void) {
  TAU_PROFILE("f1()", "", TAU_USER);
  sleep(1);
  return 0;
}


int input(void) {
  TAU_PROFILE("input", "", TAU_USER);
  sleep(1);
  return 0;
}

int output(void) {
  TAU_PROFILE("output", "", TAU_USER);
  sleep(1);
  return 0;
}

int f2(void) {
  TAU_PROFILE("f2()", "", TAU_USER);

  TAU_PHASE_CREATE_STATIC(t,"IO Phase", "", TAU_USER);
  TAU_PHASE_START(t);
  input();
  output();
  TAU_PHASE_STOP(t);
  return 0;
}


int f3(int x) {
  TAU_PROFILE("f3()", "", TAU_USER);
  sleep(x);
  return 0;
}

int f4(void) {
  TAU_PROFILE("f4()", "", TAU_USER);
  sleep(1);
  return 0;
}

int main(int argc, char **argv) {
  TAU_PROFILE("main()", "", TAU_DEFAULT);
  TAU_PROFILE_SET_NODE(0);

  for (int i=0; i<5; i++) {
    char buf[32];
    snprintf(buf, sizeof(buf),  "Iteration %d", i);

    TAU_PHASE_CREATE_DYNAMIC(phase, buf, "", TAU_USER);
    TAU_PHASE_START(phase);


    printf("Iteration %d\n", i);
    f1();
    f2();
    f3(i);
    
    TAU_PHASE_STOP(phase);
  }

  f1();
  f4();
  return 0;
}

File: examples/phase/fortran/simple.f
cc simple.f
cc --------
cc This file contains sample code for phased based profiling with TAU
cc-----------------------------------------------------------------------------


      subroutine F1()
        integer profiler(2) / 0, 0 /
        save    profiler

        call TAU_PROFILE_TIMER(profiler,'f1()')
        call TAU_PROFILE_START(profiler)
        call SLEEP(1)
        call TAU_PROFILE_STOP(profiler)
      end

      subroutine INPUT()
        integer profiler(2) / 0, 0 /
        save    profiler

        call TAU_PROFILE_TIMER(profiler,'input')
        call TAU_PROFILE_START(profiler)
        call SLEEP(1)
        call TAU_PROFILE_STOP(profiler)
      end

      subroutine OUTPUT()
        integer profiler(2) / 0, 0 /
        save    profiler

        call TAU_PROFILE_TIMER(profiler,'output')
        call TAU_PROFILE_START(profiler)
        call SLEEP(1)
        call TAU_PROFILE_STOP(profiler)
      end


      subroutine F2()
        integer profiler(2) / 0, 0 /
        save    profiler

        integer phase(2) / 0, 0 /
        save    phase

        call TAU_PROFILE_TIMER(profiler,'f2()')
        call TAU_PROFILE_START(profiler)

        call TAU_PHASE_CREATE_STATIC(phase,'IO Phase')
        call TAU_PHASE_START(phase)

        call INPUT()
        call OUTPUT()
        
        call TAU_PHASE_STOP(phase)

        call TAU_PROFILE_STOP(profiler)
      end


      subroutine F3(val)
        integer val
        integer profiler(2) / 0, 0 /
        save    profiler

        call TAU_PROFILE_TIMER(profiler,'f3()')
        call TAU_PROFILE_START(profiler)
        call SLEEP(val)
        call TAU_PROFILE_STOP(profiler)
      end


      subroutine F4()
        integer profiler(2) / 0, 0 /
        save    profiler

        call TAU_PROFILE_TIMER(profiler,'f4()')
        call TAU_PROFILE_START(profiler)
        call SLEEP(1)
        call TAU_PROFILE_STOP(profiler)
      end


      subroutine ITERATION(val)
        integer val
        character(13) cvar 
        integer profiler(2) / 0, 0 /
        save profiler

        print *, "Iteration ", val

        write (cvar,'(a9,i2)') 'Iteration', val
        call TAU_PHASE_CREATE_DYNAMIC(profiler, cvar)
        call TAU_PHASE_START(profiler)

        call F1()
        call F2()
        call F3(val)
        call TAU_PHASE_STOP(profiler)
        return
      end 

      program main
        integer i
        integer profiler(2) / 0, 0 /
        save    profiler

        call TAU_PROFILE_INIT()
        call TAU_PROFILE_TIMER(profiler, 'main()')
        call TAU_PROFILE_START(profiler)
        call TAU_PROFILE_SET_NODE(0)


        do 10, i = 0, 4
        call ITERATION(i)
10      continue


        call F1()
        call F4()
        call TAU_PROFILE_STOP(profiler)
      end

